src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs

```diff
@@ -47,16 +47,25 @@ public void BeginOperation(CancellationToken cancellationToken, Action<object?>
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            _awaitableState |= AwaitableState.Running;
-
             // Don't register if already completed, we would immediately unregistered in ObserveCancellation
             if (cancellationToken.CanBeCanceled && !IsCompleted)
             {
+                _cancellationTokenRegistration = cancellationToken.UnsafeRegister(callback, state);
+
+                // If we get back the default CancellationTokenRegistration then it means the
+                // callback synchronously and we can throw inline. This is safe because we haven't changed
+                // the state of the awaitable as yet.
+                if (_cancellationTokenRegistration == default(CancellationTokenRegistration))
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
```

Comment:
No, we need to make sure that this only happens if the registration is a default one. Otherwise, we need to dispose it before throwing, which would result in a dead lock if the token fired before we throw the exception.

Reply:
I guess it will make already canceled reads more expensive.